dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,156 @@
+offline integrity measurement servers on the USB device and clients running in
+Dynamically Launched Measured Environments (DLME). It should allow the Fobnail
+owner to verify the trustworthiness of the running system before performing any
+sensitive operation.
+
+According to the information provided at TrenchBoot Summit 2021 (the fragment
+discussed here - https://youtu.be/xZoCtNV8Qs0?t=5062), the minimum OS, together
+with the connected Fobnail Token, should allow determining the state of the
+platform before the target OS is launched. As we can see in the presentation,
+`heads` would be responsible for the platform's trust control, while in this
+document, we want to present the results of research on the possibilities of
+using various operating systems.
+
+## Different OSs propositions
+
+The research effect is presented below. 4 systems were considered:
+
+* Zephyr
+* Xous
+* seL4
+* Linux
+
+Each of them has a short description, an analysis of the launch in DLME, and the
+possibilities and potential problems that will have to be addressed for the
+Fobnail Token to be functional.
+
+### Zephyr
+
+Zephyr is a well-known and supported embedded RTOS with a monolithic kernel
+written in C. It supports various devices ranging from Cortex-M to x86 CPUs and
+provides support for IPv4, CoAP, and USB needed for communication with Fobnail
+Token.
+
+#### Running in DLME
+
+Zephyr boots over Multiboot1, but SKL supports only Muliboot2, so either SKL
+must be extended to gain Multiboot1 support or Zephyr to gain Multiboot2. Also,
+Zephyr is not portable - some important configuration is baked into Zephyr
+during build, like LAPIC base address - the base address is set by
+`CONFIG_LOAPIC_BASE_ADDRESS`, if `IA32_APIC_BASE` MSR doesn't match, Zephyr will
+break. The same thing for xAPIC vs x2APIC mode. If Zephyr was configured to use
+xAPIC mode, it might not boot on a platform with x2APIC enabled in its firmware.
+
+#### Fobnail integration
+
+Zephyr provides most of the drivers needed for integration:
+
+- USB hub drivers
+- USB EEM driver
+- Network stack with IPv4 and CoAP support
+
+TPM driver is missing, however there is a fairly recent
+[PoC implementation](https://github.com/drandreas/zephyr-tpm2-poc) of TPM2
+stack.
+
+Zephyr provides good support for standard C library (newlib), which should
+simplify [fobnail-attester](https://github.com/fobnail/fobnail-attester)
+porting.
+
+### Xous
+
+Xous is a microkernel-based OS written in Rust, currently in Alpha state. Xous
+is used as a secure microkernel for [Betrusted](betrusted.io) device and is
+focused on RISC-V, which right now is the only architecture supported. However,
+x86 already contains
+[dummy implementation](https://github.com/betrusted-io/xous-core/blob/30b82b25b100e958790973c129dc49e1acca79ec/kernel/src/arch/x86_64.rs)
+which probably will be extended someday.
+
+
+#### Running in DLME
+
+Xous currently cannot run on x86, and DLME on RISC-V is not covered here.
```

Comment:
@arturkow2000 So there is no way to run RISC-V arch in DLME?

Reply:
@TomaszAIR Probably it is but I don't know how to do it.

Reply:
@arturkow2000 we could submit an issue in Xous community as mentioned here https://github.com/fobnail/docs/pull/26#discussion_r855083920

Reply:
issue submitted https://github.com/betrusted-io/xous-core/issues/139